.github/workflows/update-gemfiles.yml

```diff
@@ -0,0 +1,63 @@
+name: Update Gemfiles
+on:
+  # Execute on `push` and not `pull_request` because `pull_request`
+  # always compares if the `paths` have changed compared to the PR base.
+  # This is an issue because it means that all commits to the branch
+  # will trigger the gemfile update process, which is unnecessary and expensive.
+  #
+  # By executing on `push`, GitHub compares `paths` with the parent commit,
+  # meaning the gemfile update process will only execute on the exact commit
+  # that changes any of the `paths`.
+  #
+  # Because this process is slow and expensive, and we commit the gemfile changes back
+  # to the branch, we have an additional filter to only execute this action on branches
+  # attached to a PR.
+  #
+  # We could do the inverse: execute this action on `pull_request`, and additionally check
+  # if `paths` was changed compared to the parent commit, but this proved more complicated.
+  push:
+    paths:
+      # Files that declare the dependency tree
+      - Gemfile
+      - Appraisals
+      - ddtrace.gemspec
+      # Files that control gemfile generation
+      - tasks/appraisal.rake
+      - .github/workflows/update-gemfiles.yml
+      # In case the generated files were updated manually or in a merge commit
+      - appraisal
+      - gemfiles
+
+# Ensure obsolete job is cancelled if another commit is pushed to the same branch.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  check:
+    name: Update Gemfiles
+    runs-on: ubuntu-22.04
+    steps:
+      # Only execute if there's a PR attached to this branch.
+      # Because we execute on `push`, we have to double check here if this is part of a PR.
+      - name: Check if this branch is attached to a Pull Request
+        uses: 8bitjonny/gh-get-current-pr@2215326c76d51bfa3f2af0a470f32677f6c0cae9 # v2.2.0
+        id: pr
+      - if: steps.pr.outputs.pr_found == 'true'
+        uses: actions/checkout@v4
+      - if: steps.pr.outputs.pr_found == 'true'
+        uses: ruby/setup-ruby@31a7f6d628878b80bc63375a93ae079ec50a1601 # v1.143.0
+        with:
+          ruby-version: '3.2'
+          bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+      - if: steps.pr.outputs.pr_found == 'true'
+        name: Update gemfiles/*.gemfile
+        run: bundle exec rake appraisal:generate
+      - if: steps.pr.outputs.pr_found == 'true'
+        name: Update gemfiles/*.gemfile.lock
+        run: bundle exec rake appraisal:lock
+      - if: steps.pr.outputs.pr_found == 'true'
+        name: Commit gemfiles changes, if any, back to the branch
+        uses: stefanzweifel/git-auto-commit-action@3ea6ae190baf489ba007f7c92608f33ce20ef04a # v4.16.0
+        with:
+          commit_message: Update gemfiles/*
```

Comment:
I'm not super familiar with git workflows, so maybe I'm just missing it, but is this workflow required to complete before being able to merge to master? I think that's reasonably important, so we don't wind up commiting gemfile changes to a branch that's already been merged, and thus it never makes it to master.

Reply:
That's a good point. To enable a mandatory GitHub check the job should be able to always succeed.
The way it is currently implemented, it won't run on PRs that don't change gem files, thus the check won't pass.

I'll change it so that it runs (and passes) if no changes to gems have taken place.

Reply:
Fixed it, thank you!